plugin/nginx_plus_collectd.py

```diff
@@ -772,6 +834,42 @@ def get_processes(self):
         '''
         return self._send_get(self.processes_url)
 
+    def _get_api_version(self):
+        '''
+        Determines whether the Nginx-plus plugin has a versioned API or legacy API.
+        '''
+        newer_api_base_path = self.api_base_path if self.api_base_path is not None else '/api'
+        legacy_api_base_path = self.api_base_path if self.api_base_path is not None else '/status'
+        base_url = 'http://{}:{}'.format(self.status_host, str(self.status_port))
+
+        try:
+            response = requests.get("{}{}/{}".format(base_url, newer_api_base_path, DEFAULT_API_VERSION), auth=self.auth_tuple)
+            if response.status_code == requests.codes.ok:
+                return DEFAULT_API_VERSION
+            else:
+                response = requests.get("{}{}".format(base_url, legacy_api_base_path), auth=self.auth_tuple)
+                if response.status_code == requests.codes.ok:
+                    return 0
+            LOGGER.error(
+                "Failed to detect the Nginx-plus API type (versioned or legacy), please check your input configuration.")
+        except RequestException as e:
+            LOGGER.exception("Failed to detect the Nginx-plus API type (versioned or legacy), due to the error: %s", e)
+        return None
+
+    def validate_nginx_version(self):
```

Comment:
What is the benefit of this function? Are metrics suddenly inaccessible if an upgrade completes?

Reply:
In case of upgrades like versioned to versioned (r17 -> r25) or legacy to legacy (r10 -> r12), there will be no issue but in case of legacy to versioned or versioned to legacy our code will break as all the endpoints will be different. So, we are just exiting the code when there is any change in the version while the plugin is running.

Reply:
Would version detection be possible if no version or base path was configured?  Ideally we'd be able to handle this assuming existing configurations allow it.  Also I think these could both be `RuntimeError` if we aren't using custom types since `ValueError` is generally for function argument issues.

Reply:
> Would version detection be possible if no version or base path was configured?  

Yes, It would be possible to detect NGINX+ API type (versioned or legacy) using the default values, if `APIVersion` and/or `APIBasePath` are not configured.

> Also I think these could both be RuntimeError if we aren't using custom types since ValueError is generally for function argument issues.  

We have changed the exception type to `RuntimeError`.